README


Ingres Corporation



Ingres Ruby Active Record Adapter/Driver Version 1.3.1

-------------------------------------------------------------------------------

  1. Welcome
  2. Overview

       1. New_in_This_Release

  3. Operating_System_Support
  4. Installation_Considerations
  5. General_Considerations

       1. Features_Not_Included
       2. Syntax_for_the_Ingres_Ruby_Driver
       3. Syntax_for_the_Ingres_AR_Adapter

  6. Building_and_Installing_the_Ingres_Ruby_AR_Adapter/Driver

       1. Ingres_Ruby_Source_Code
       2. Ingres_Ruby_Driver
       3. Ingres_AR_Adapter

  7. Example_Code

       1. IRB_(Interactive_Ruby)_Using_Ingres_Ruby_Driver_Only_
          (Without_AR_Adapter)
       2. Ruby_Program_Using_Ingres_Ruby_Driver_Only_(Without_AR_Adapter)
       3. Rails_Cookbook_Tutorial_(Web_Application)_Using_Ingres_AR_Adapter
       4. Ingres_Flight_Planner_Demo_Using_Ingres_Ruby_Driver_Only_
          (Without_AR_Adapter)

  8. Known_Issues
  9. More_Information

-------------------------------------------------------------------------------

Welcome

This readme contains all of the documentation on the Ingres Ruby Active Record
(AR) adapter and driver. AR is the standard database access interface used by
Ruby on Rails (ROR or Rails) applications.
Please review this readme before building or installing this software. We
encourage users to test the software and provide feedback.
-------------------------------------------------------------------------------

Overview

There are two components to the Ingres support for Ruby:

* Ingres Active Record adapter
* Ingres Ruby driver

Both components are required for Rails applications using Active Record (AR),
which is the standard Rails Object-to-Relations Mapping (ORM) interface. The
adapter implements the AR classes for Ingres; internally, it invokes the
classes and methods in the driver to actually communicate with Ingres.
The driver, while primarily intended to service the adapter, can also be
invoked directly from any Ruby program to communicate with Ingres. The driver
interface does not follow any industry standard and may change in the future,
if required.
To summarize, the AR interface provides an object interface while the driver
interface is a more direct SQL-level interface.
-------------------------------------------------------------------------------

New in This Release

This is a minor release that addresses the following:

* Ruby on Rails adapter

  o Add support for the correct null clause for Ingres 9.1.1 and earlier


-------------------------------------------------------------------------------

Operating System Support

This Ingres Ruby Active Record adapter/driver supports all of the platforms
supported by Ingres, including:

* Solaris
* HP-UX
* AIX
* Linux
* Windows

-------------------------------------------------------------------------------

Installation Considerations

To build and install the Ingres Ruby interface, the following components are
required:

* Ingres 2.6 or above. At a minimum, an Ingres client installation is required
  on the same machine as the Ruby installation. For a list of Ingres binary and
  source downloads, see http://www.ingres.com.
* Ruby, preferably at version 1.8.5 or later. Additional Ruby components or
  related products (such as Active Record or Rails) may also be required
  depending on how Ruby will be used with Ingres. See below for details.
* C compiler (for example, GNU/C on Linux and UNIX, or Microsoft Visual Studio
  6 on Windows). The C compiler is not needed if using a pre-built version of
  the Ingres Ruby driver.
  Note: Ruby does not yet support building on later versions of Microsoft
  Visual Studio.
* The Ingres Ruby AR adapter source code and driver binary. The source code for
  the driver is also needed if not using the pre-built binary.

-------------------------------------------------------------------------------

General Considerations

-------------------------------------------------------------------------------

Features Not Included

The following features are currently not included in the Ingres Open Source
Ruby interface (also see the Known Issues section):

* Only one Ingres connection at a time is supported from an application.

-------------------------------------------------------------------------------

Syntax for the Ingres Ruby Driver

Note: We do not recommend that production applications use this API since it is
intended for the Ingres AR adapter. The API may change in the future. It is,
however, useful for testing Ingres connectivity and functionality.
Ruby program must have a "require 'Ingres'" statement to load in driver file.
Create Ingres class object:

  object_instance_name=Ingres.new

Methods (by category):
Connect to database:

  connect(database_name)

  connect_with_credentials(database_name, user, password)

Disconnect from database:

  disconnect()

Execute SQL queries and transactional statements:

  execute(sql)

(Alias for "execute" is "exec".)
For SQL queries or database procedure calls that contain parameters, use the
pexecute() method in one of the following formats.
For selects, inserts, deletes, and updates:

  pexecute(sql [ [, param_type, param_value ] … ] )

For database procedure calls:

  pexecute("{ <call> | <execute procedure> procedure_name [ (column_name = ? [
  [ , column_name = ? ] … ] ) ] }" [ [ , column_name, param_type, param_value
  ] … ] )

where parameters in the sql text are represented by question marks (?) and
there is one set of param_type/param_value entries for each one.
param_type must be one of the following single characters:
 _______________________
|param_type|Description_|
|b_________|byte________|
|B_________|long_byte___|
|c_________|char________|
|d_________|date,_______|
|D_________|decimal_____|
|f_________|float_______|
|i_________|integer_____|
|n_________|nchar_______|
|N_________|nvarchar____|
|t_________|text________|
|T_________|long_text___|
|v_________|varchar_____|
|V_________|long_varchar|

param_value should correspond to the type.
Both methods return a result set consisting of an array of rows, each of which
is in turn an array of the columns within the row. For example:

  ing.pexecute("select * from t1 where f1 = ?", 'i', 2)

Result Set Metadata:
 ____________________________________________________________________________
|Method________________|Description__________________________________________|
|rows_affected()_______|Returns_the_number_of_rows_affected_by_last_execute()|
|column_list_of_names()|Returns_the_names_of_the_columns_in_the_result_set___|
|data_types()          |Returns the data types of the columns in the result  |
|______________________|set__________________________________________________|
|data_sizes()          |Returns the data lengths of the columns in the result|
|______________________|set__________________________________________________|

Database Metadata:
 _____________________________________________________________________
|Method____________|Description_______________________________________|
|current_database()|Returns_name_of_the_current_database_connected_to_|
|tables()__________|Returns_list_of_all_tables_in_the_current_database|

Miscellaneous:
To turn trace debugging on or off:

  set_debug_flag(flag, flag_value)

where
flag is one of following:

* GLOBAL_DEBUG
* DEBUG_TRANSACTIONS
* DEBUG_SQL
* DEBUG_TERMINATION

flag_value is TRUE or FALSE.
-------------------------------------------------------------------------------

Syntax for the Ingres AR Adapter

See Ruby Active Record documentation: http://ar.rubyonrails.com/.

ActiveRecord Type Mappings

The following table shows the type mappings between Ingres, Ruby and
ActiveRecord.
 _____________________________________________________________________________
|ActiveRecord_Type|Ruby_Class|Ingres_Type_______________|Notes________________|
|:primary_key     |          |INTEGER NOT NULL PRIMARY  |                     |
|_________________|__________|KEY_______________________|_____________________|
|:string          |String    |CHAR, NCHAR, VARCHAR,     |                     |
|_________________|__________|NVARCHAR,INTERVAL_________|_____________________|
|:text____________|String____|VARCHAR(32000)____________|_____________________|
|:boolean_________|Boolean___|INTEGER1__________________|_____________________|
|:integer_________|Fixnum____|INTEGER2|4|8______________|_____________________|
|                 |Bignum    |if INTEGER4|8 won't fit in|                     |
|_________________|__________|Fixnum____________________|_____________________|
|:float___________|Float_____|FLOAT4,_FLOAT8,_MONEY_____|_____________________|
|:decimal_________|String____|DECIMAL___________________|_____________________|
|:date____________|Date______|ANSIDATE__________________|Ingres_9.1.0_or_later|
|:datetime________|Time______|DATE,_INGRESDATE__________|_____________________|
|:time____________|Time______|TIME______________________|Ingres_9.1.0_or_later|
|:timestamp_______|Time______|TIMESTAMP_________________|Ingres_9.1.0_or_later|

Note that the :binary ActiveRecord type is missing due to the lack of support
for Ingres LONG BYTE, LONG VARCHAR support.
-------------------------------------------------------------------------------

Building and Installing the Ingres Ruby AR Adapter/Driver

-------------------------------------------------------------------------------

Ingres Ruby Source Code

The source code for the Ingres Ruby driver is kept in a Subversion repository
at http://code.ingres.com. In order to download the source you will need an SVN
client. Below is a list of some of the clients available for working with
Subversion

* Windows - TortoiseSVN - http://tortoisesvn.tigris.org/
* Linux - "subversion" package for your distribution
* Unix - Build from sources http://subversion.tigris.org/
* OS X

  o Subversion from http://www.macports.org/
  o Versions from http://www.versionsapp.com/


To download the source you need to specify the following URL when performing a
checkout:

* Ruby - http://code.ingres.com/ingres/drivers/ruby/

For example if using the command line Subversion client 'svn' the following
will check out all the code for the Ingres Ruby driver:

  svn co http://code.ingres.com/ingres/drivers/ruby/

This will download the complete tree for the Ruby driver into the directory
'ruby'. If you are only interested in the 'main' code line use the following
command:

  svn co http://code.ingres.com/ingres/drivers/ruby/main/ ruby-main

Note that an additional parameter of 'ruby-main' has been specified. The code
for http://code.ingres.com/ingres/drivers/ruby/main/ will be extracted in to
this directory. Otherwise the checkout will download the code into a directory
named 'main'. For more information on working with Ingres source code control
system see - http://community.ingres.com/wiki/Connectivity_Drivers_Source_Code
-------------------------------------------------------------------------------

Ingres Ruby Driver

A binary version of the driver is provided for Windows (Ingres.so). If using
the provided binary, skip to the install step below.
The driver can also be built from the source provided (Ingres.c). To build and
install the Ingres Ruby driver, the following components are needed AND MUST BE
IN YOUR PATH:

* Ingres installation
* C compiler (for example, GNU/C on Linux and UNIX, or Microsoft Visual Studio
  6 on Windows).

To build the driver

  1. Copy source Ingres.c and extconf.rb to your build directory (anywhere).
  2. Create the makefile using the Ruby mkmf utility. Run the following command
     after customizing it to your installation of Ingres:

       ruby -r mkmf extconf.rb
           --with-ingres-include='/opt/Ingres/IngresII/ingres/files/'
           --with-ingres-lib='/opt/Ingres/IngresII/ingres/lib/'

     A sample createMake.bat has been included for Windows. The utility mkmf
     pulls configuration information from the extconf.rb file, which may need
     slight library name changes for different versions of Ingres or operating
     systems; see comments in extconf.rb file.
     The output should look similar to the following:

       bash-2.05$ ruby -r mkmf extconf.rb --with-ingres-include='/usr/ingres/
       files/' --with-ingres-lib='/usr/ingres/lib'
       checking for iiapi.h... yes
       checking for main() in -lingres... yes
       checking for main() in -lingres... yes
       creating Makefile

  3. Compile and link the driver. Type make on Linux and UNIX or nmake on
     Windows. This should compile and link cleanly, creating output shared
     library Ingres.so.

     o On Windows, this is a DLL, but is nevertheless suffixed with ".so"
       rather than ".dll".
     o On Windows, you can ignore warnings about the compiler and linker flags
       being unknown options.
     o On Windows with Ruby versions later than 1.8.5, the following error may
       be seen on the link:

         cannot open input file "msvcrt-ruby18.lib"

       This can be resolved by removing the (lower case) "libpath =" line in
       the makefile.


To install the driver
Type make install. (On Windows, type nmake).
Or, with the pre-built Windows binary, copy Ingres.so to Ruby i386-msvcrt
directory, such as: c:\ruby\lib\ruby\site_ruby\1.8\i386-msvcrt.
-------------------------------------------------------------------------------

Ingres AR Adapter

At a minimum, Ruby's Active Record must be installed. If running Rails, this
will already exist as part of the Rails installation. For non-Rails users, AR
may come with Ruby or may be installed separately as a Ruby "gem" into the Ruby
installation.
There are several steps required to install the Ingres AR adapter into the Ruby
AR installation.
Note: If only using the Ingres Ruby driver interface (not AR), installation of
the adapter can be skipped and you can go directly to installation of the
driver below.
Typically, the base path of the AR installation will be similar to the
following and will be referred to hereafter in this documentation as AR_BASE
(where the last digits in the path refer to the AR version):
UNIX/Linux: /usr/local/lib/ruby/gems/1.8/gems/activerecord-1.14.2
Windows: C:\ruby\lib\ruby\gems\1.8\gems\activerecord-1.15.3
To install the Ingres AR adapter into the Ruby AR installation

  1. Copy the adapter file, ingres_adapter.rb, to the AR_BASE/lib/active-
     record/connection_adapters directory.
  2. NOTE: Not required for Rails 2.1 or later Add Ingres as a valid database
     to AR. Edit active_record.rb file in AR_BASE/lib by adding "ingres" to the
     end of a line that looks like:

       RAILS_CONNECTION_ADAPTERS = %w( mysql postgresql sqlite firebird
       sqlserver db2 oracle sybase openbase ingres )

  3. Set up database connection information. Create a new directory called
     native_ingres under the AR_BASE/test/connections directory. Copy and edit
     the sample connection.rb file to define database name, user ID, password
     information.
  4. Set II_DATE_FORMAT to SWEDEN. Either set this as an OS environment
     variable in your RAILS application environment or as an Ingres variable
     (with the ingsetenv command). This restriction will be removed in the
     future. Also, code in the adapter file, if uncommented (scan for SWEDEN),
     may allow successful processing without setting II_DATE_FORMAT.
  5. For each Rails application accessing Ingres, edit the database.yml file in
     the application's config folder to specify the Ingres adapter plus
     database and user information. The following is a sample entry for a local
     Ingres database connection named my_rails_app_development:

       development:
           adapter: ingres
           database: my_rails_app_development
           username: ingres
           password: ingres
           host: localhost

     For remote connections, change database and host to Ingres remote
     connection format in either "vnode::databasename" where vnode is a name
     configured in the Ingres Network Utility or a dynamic vnode in
     @host,protocol,port[user,pwd]" format. For details, see the Ingres
     Connectivity Guide.

-------------------------------------------------------------------------------

Example Code

-------------------------------------------------------------------------------

IRB (Interactive Ruby) Using Ingres Ruby Driver Only (Without AR Adapter)


  >>>$ irb

  irb(main):001:0> require 'Ingres'

  => true

  irb(main):002:0> i = Ingres.new

  => #<Ingres:0xb7fba520>

  irb(main):003:0> i.connect("activerecord_unittest")

  => #<Ingres:0xb7fba520>

  irb(main):006:0> sql="select * from developers"

  => "select * from developers"

  irb(main):007:0> result_set = i.execute(sql)

  => [[5, "fixture_5", 100000, "NULL", "NULL"], [6, "fixture_6", 100000,
  "NULL", "NULL"], [7,"fixture_7", 100000, "NULL", "NULL"], [11, "Jamis", 9000,
  "NULL", "NULL"], [8, "fixture_8", 100000, "NULL", "NULL"], [9, "fixture_9",
  100000, "NULL", "NULL"], [2, "Jamis", 150000, "NULL", "NULL"], [10,
  "fixture_10", 100000, "NULL", "NULL"], [3, "fixture_3", 100000, "NULL",
  "NULL"], [1, "David", 80000, "NULL", "NULL"], [4, "fixture_4", 100000,
  "NULL", "NULL"]]

-------------------------------------------------------------------------------

Ruby Program Using Ingres Ruby Driver Only (Without AR Adapter)


  require "Ingres"
  ing = Ingres.new
  # You must connect to a database.
  # Be sure you've created the database using "createdb".
  ing.connect(some_database)
  # you can also use:
  # ing.connect_with_credentials( some_database, user_name, password)
  sql = "select * from some_table"
  result_set = ing.execute(sql)
  # or
  # result_set = ing.exec(sql)
  # the result set is an array of arrays.
  # you get one array for each row
  # {{a,b,c}, {d,e,f}, etc}
  # metadata is stored in these routines
  # how many rows were touched by the last query
  # usually an update
  puts ing.rows_affected
  # what's the current database (from the connect call)
  puts ing.current_database
  # the names of each column in the result set
  puts ing.column_list_of_names
  # the size of the data in each column
  # this could be the length of a varchar or an int
  puts ing.data_sizes
  # what type is in each column
  puts ing.data_types
  # all the tables in the current database
  puts ing.tables
  # properly terminate your database connection
  ing.disconnect

-------------------------------------------------------------------------------

Rails Cookbook Tutorial (Web Application) Using Ingres AR Adapter

A Cookbook Tutorial can be accessed from the Rails website (http://
rubyonrails.org/ docs) under the Tutorials section. Selection Rolling with Ruby
on Rails: Part I takes you to http://www.onlamp.com/pub/a/onlamp/2006/12/14/
revisiting-ruby-on-rails-revisited.html. The demo referred to in the tutorial
is installed as part of some of the prepackaged Ruby on Rails versions, such as
InstantRails on Windows or Rails LiveCD on Linux.
To run this demo in InstantRails, which is preconfigured to run against another
database, do the following:

  1. Install the Ingres Ruby AR adapter and driver into InstantRails per the
     instructions above (InstantRails comes with its own version of Ruby and
     Active Record).
  2. Create the database in Ingres with schemas.

       1. Create the database:

            createdb cookbook_development

       2. Create and populate the tables:

            sql cookbook_development < cookbook_ingres.sql


  3. Configure the application to use the Ingres ActiveRecord adapter.
     Edit database.yml in ...\rails_apps\cookbook:

       1. Change adapter from "mysql" to "ingres" in all (three) occurrences.
       2. Update username and password as needed; for example, username:
          ingres.
       3. Edit database to be the full name of the database. Unlike MySql, the
          _development, _test, _production suffixes are not automatically
          added.

  4. Start up systems (if not already started):

       1. Start up Ingres (ingstart)
       2. Start up Rails. For Instant Rails:

            cd d:\InstantRails
            use_ruby
            cd ..
            InstantRails

       3. Start up Mongrel (web server)
          If Mongrel is already up and app or even database.yml was changed,
          you must restart Mongrel:
          Shutdown with Control-C in Mongrel startup command window.

            cd d:\InstantRails\rails_apps\cookbook
            mongrel_rails start


  5. Run application:
     From web browser, set url=http://localhost:3000/category. You should get a
     screen showing "Online Cookbook" at the top.

-------------------------------------------------------------------------------

Ingres Flight Planner Demo Using Ingres Ruby Driver Only (Without AR Adapter)

If you would like to get a copy of this, please post a request on the Ingres
Community Forum in the Database Drivers and APIs area.
-------------------------------------------------------------------------------

Known Issues

Known issues are as follows:

* Errors (such as trying to read an unsupported data type) may leave connection
  state in an unusable state, causing errors such as "The requested operaton
  cannot be performed with active transactions."
* Driver will not build on later versions of Microsoft Visual Studio (2003 and

CHANGELOG
version 1.4.1
- Fix #656 - getProcedureName() returns a non-NULL terminated procedure name
- Fix #653 - Update s.date dynamically for each release
- Fix #651 - Add support for building a Ruby Gem
- Fix #555 - SEGV when fetcing LOB data
- Fix #565 - Compilation errors with Redhat ES 5.4 (x86-64)

version 1.4.0
- Default to "id" for the :primary_key column in ActiveRecord if no 
  PRIMARY KEY constraint is defined
- Update the documentation using RDoc
- Add Ingres.set_environment()
- Add the ability to set the date format for ingresdate values at connection
  time
- Update ingres_adapter.rb for ActiveRecord 2.3.x
- Fix #489 - librt.so should only be used on Linux
- Use IIAPI_VCH_TYPE for converting date values to their string form
- Add support for the SQL SAVEPOINT statement
- Fix a regression where certain query would return no data
- Add support for multiple active connections
- replaced all calls to ii_allocate()/ii_reallocate() with their ruby
  counterparts ALLOC_N and REALLOC_N
  - Except for memory allocation in getColumn()
- Added a SQL query classifier to eliminate the need to force the query into
  upper case
- Moved Unicode functionality into Unicode.c and Unicode.h
- Moved defines/typedefs/function prototypes into Ingres.h
- Added new unit tests directory, see "Testing Done" for more information
- Ingres.connect() takes an optional username/password
  - eliminates the need for connect_with_credentials (now an alias pointing to
    connect)
- Ingres.execute() now accepts a variable number of parameters, requires query
  but parameter values is optional
  - Ingres.pexecute() now an alias for Ingres.execute()
- Re-formatted the code using astyle (http://astyle.sourceforge.net/)
- Added the following Ruby constants
  - VERSION - version string, e.g. "1.4.0-dev"
  - API_LEVEL - value for IIAPI_VERSION
  - REVISION - the SVN revision for the driver
- CM_AttrTab should be defined as an extern
- Improve Ingres library and header detection _ version 1.3.0
- Add support for the null clause used in Ingres 9.1.x (and earlier)
- Change status to Beta

version 1.3.0
- Add quote_column_names() for quoting of table and column names. Required
  for the ActiveRecord base unit tests
- Fix data truncation for the ActiveRecord :text type
- Add support for Ingres datatypes float4, float8, money, decimal, bigint,
  and ANSI date/time/timestamp.
- Add support in ingres_adapter.rb for Rails migrations
- Restrict tables() to non-system tables

version 1.2.2
- Add an ascii README file for systems without access to a browser
- Add TODO.txt show show areas that require attention
- Add missing GPL header text to ingres_adapter.rb
- Fix quote() in ingres_adapater.rb to use the class of variable rather than
  the equivalent Ingres column type.
- Fix segvio in Ingres Ruby driver when fetching char/varchar with data
  longer than 4074 bytes.

version 1.2.1
- Initial release

/*
vim: ts=4 sw=4 expandtab textwidth=78
$Id$
*/


ext/Ingres.h
/*
**      Copyright (c) 2009 Ingres Corporation
**
**      This program is free software; you can redistribute it and/or modify
**      it under the terms of the GNU General Public License version 2 as
**      published by the Free Software Foundation.
**
**      This program is distributed in the hope that it will be useful,
**      but WITHOUT ANY WARRANTY; without even the implied warranty of
**      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**      GNU General Public License for more details.
**
**      You should have received a copy of the GNU General Public License along
**      with this program; if not, write to the Free Software Foundation, Inc.,
**      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

/*      static char     Sccsid[] = "$Id$";        */

/*
**      INGRES.H
**
**      History
**              08/24/2009 (grant.croker@ingres.com)
**                  Created
**              11/09/2009 (grant.croker@ingres.com)
**                  Add the ability to set the date format for Ingres date values
**                  from within Ruby 
**              03/30/2010 (grant.croker@ingres.com)
**                      Fix #565 - Compilation errors with Redhat ES 5.4 (x86-64)
**              04/07/2010 (grant.croker@ingres.com)
**                      Add missing #include for version.h
*/

#include <version.h>

/* Constants */
#define RUBY_INGRES_VERSION "1.4.1"
#define RUBY_INGRES_API IIAPI_VERSION

#define RUBY_STRING    			"STRING"
#define RUBY_INTEGER   			"INTEGER"
#define RUBY_BYTE      			"BYTE"
#define RUBY_TEXT      			"TEXT"
#define RUBY_VARCHAR   			"VARCHAR"
#define RUBY_DATE      			"DATE"
#define RUBY_DOUBLE    			"DOUBLE"
#define RUBY_TINYINT   			"TINYINT"
#define RUBY_LOB       			"LOB"
#define RUBY_UNMAPPED  			"UNMAPPED_DATATYPE"
#define ERROR_MESSAGE_HEADER 		"Ingres Diagnostic Messages:\n"
#define RubyParamOffset(ip, ipc) 	((ip - ipc) * (2 + ipc))

#define RUBY_NVARCHAR_PARAMETER		'N'
#define RUBY_NCHAR_PARAMETER		'n'
#define RUBY_VARCHAR_PARAMETER		'v'
#define RUBY_DECIMAL_PARAMETER		'D'
#define RUBY_LONG_BYTE_PARAMETER	'B'
#define RUBY_LONG_TEXT_PARAMETER	'T'
#define RUBY_LONG_VARCHAR_PARAMETER	'V'

#define RUBY_INTEGER_PARAMETER		'i'
#define RUBY_BYTE_PARAMETER		'b'
#define RUBY_CHAR_PARAMETER		'c'
#define RUBY_DATE_PARAMETER		'd'
#define RUBY_TEXT_PARAMETER		't'
#define RUBY_FLOAT_PARAMETER		'f'

#define DECIMAL_BUFFER_LEN		16
#define DECIMAL_PRECISION		31
#define DECIMAL_SCALE			15
#define DOUBLE_PRECISION		31
#define DOUBLE_SCALE			15
#define MAX_CHAR_SIZE		        32000  /* max #bytes Ingres (var)char */
#define LOB_SEGMENT_SIZE 8192

/* Ingres 2.6 is missing a define for IIAPI_CPV_DFRMT_ISO4 */
#if !defined(IIAPI_CPV_DFRMT_ISO4)
  #define IIAPI_CPV_DFRMT_ISO4 9
#endif

/* Query types */

#define INGRES_SQL_SELECT             0
#define INGRES_SQL_INSERT             1
#define INGRES_SQL_UPDATE             2
#define INGRES_SQL_DELETE             3
#define INGRES_SQL_COMMIT             4
#define INGRES_SQL_ROLLBACK           5
#define INGRES_SQL_OPEN               6
#define INGRES_SQL_CLOSE              7
#define INGRES_SQL_CONNECT            8
#define INGRES_SQL_DISCONNECT         9 
#define INGRES_SQL_GETDBEVENT         10
#define INGRES_SQL_SAVEPOINT          11
#define INGRES_SQL_AUTOCOMMIT         12
#define INGRES_SQL_EXECUTE_PROCEDURE  13
#define INGRES_SQL_CALL               14
#define INGRES_SQL_COPY               15
#define INGRES_SQL_CREATE             16
#define INGRES_SQL_ALTER              17
#define INGRES_SQL_DROP               18
#define INGRES_SQL_GRANT              19
#define INGRES_SQL_REVOKE             20
#define INGRES_SQL_MODIFY             21
#define INGRES_SQL_SET                22
/* Not really an SQL Statement but INGRES_START_TRANSACTION is expected to be handled as one */
#define INGRES_START_TRANSACTION      23
#define INGRES_SQL_ROLLBACK_WORK_TO   24
#define INGRES_SQL_ROLLBACK_TO        25

#define INGRES_NO_OF_COMMANDS         26
static struct
{
  char        *command;
  int         code;
  int         length ;
} SQL_COMMANDS [INGRES_NO_OF_COMMANDS] =
{
  { "SELECT", INGRES_SQL_SELECT, 6 },
  { "INSERT", INGRES_SQL_INSERT, 6 },
  { "UPDATE", INGRES_SQL_UPDATE, 6 },
  { "DELETE", INGRES_SQL_DELETE, 6 },
  { "COMMIT", INGRES_SQL_COMMIT, 6 },
  { "ROLLBACK", INGRES_SQL_ROLLBACK, 8 },
  { "OPEN", INGRES_SQL_OPEN, 4 },
  { "CLOSE", INGRES_SQL_CLOSE, 5 },
  { "CONNECT", INGRES_SQL_CONNECT, 7 },
  { "DISCONNECT", INGRES_SQL_DISCONNECT, 10 },
  { "GET DBEVENT", INGRES_SQL_GETDBEVENT, 11 },
  { "SAVEPOINT", INGRES_SQL_SAVEPOINT, 9 },
  { "SET AUTOCOMMIT", INGRES_SQL_AUTOCOMMIT, 14 },
  { "EXECUTE PROCEDURE", INGRES_SQL_EXECUTE_PROCEDURE, 17 },
  { "CALL", INGRES_SQL_CALL, 4 },
  { "COPY", INGRES_SQL_COPY, 4 },
  { "CREATE", INGRES_SQL_CREATE, 6 },
  { "ALTER", INGRES_SQL_ALTER, 5 },
  { "DROP", INGRES_SQL_DROP, 4 },
  { "GRANT", INGRES_SQL_GRANT, 5 },
  { "REVOKE", INGRES_SQL_REVOKE, 6 },
  { "MODIFY", INGRES_SQL_MODIFY, 6 },
  { "SET", INGRES_SQL_SET, 3 },
  { "START TRANSACTION", INGRES_START_TRANSACTION, 17 },
  { "ROLLBACK WORK TO", INGRES_SQL_ROLLBACK_WORK_TO, 16 },
  { "ROLLBACK TO", INGRES_SQL_ROLLBACK_TO, 11 },
};

#define INGRES_NO_CONN_PARAMS  1
static struct
{
  char *paramName;
  II_LONG paramID;
} CONN_PARAMS [INGRES_NO_CONN_PARAMS] =
{
  {"date_format", IIAPI_CP_DATE_FORMAT },  /* for SELECT VARCHAR(date_col) */
};

#define INGRES_NO_ENV_PARAMS  1
static struct
{
  char *paramName;
  II_LONG paramID;
} ENV_PARAMS [INGRES_NO_ENV_PARAMS] =
{
  {"date_format", IIAPI_EP_DATE_FORMAT },  /* for SELECT date_col */
};

/* Cursor Modes */
#define INGRES_CURSOR_READONLY 0
#define INGRES_CURSOR_UPDATE 1

typedef struct _II_GLOBALS
{
  II_PTR envHandle;
  int debug;
  int debug_connection;
  int debug_sql;
  int debug_termination;
  int debug_transactions;
} II_GLOBALS;

typedef struct _II_SAVEPOINT_ENTRY
{
  char   *savePtName;
  II_PTR savePtHandle;
  II_PTR nextSavePtEntry;
} II_SAVEPOINT_ENTRY;

typedef struct _II_CONN
{
  int autocommit;
  II_PTR connHandle;
  II_PTR tranHandle;
  II_PTR stmtHandle;
  II_PTR envHandle;
  II_LONG fieldCount;
  II_LONG lobSegmentSize;
  IIAPI_DESCRIPTOR *descriptor;
  II_CHAR *errorText;
  II_CHAR sqlstate[6];
  II_LONG errorCode;
  int apiLevel;
  int paramCount;
  int allow_persistent;
  int num_persistent;
  char *cursor_id;
  long cursor_mode;
  char *currentDatabase;
  int queryType;
  VALUE keep_me;
  VALUE resultset;
  VALUE r_column_names;
  VALUE r_data_sizes;
  VALUE r_data_types;
  II_LONG columnCount;
  II_PTR savePtList;
  II_SAVEPOINT_ENTRY *lastSavePtEntry;/* Pointer to the last savePtEntry on savePtList */
} II_CONN;

typedef struct _RUBY_IIAPI_DATAVALUE
{
  IIAPI_DATAVALUE dataValue[1];
  long dv_length;
} RUBY_IIAPI_DATAVALUE;

typedef struct _RUBY_PARAMETER
{
  VALUE vkey;
  VALUE vtype;
  VALUE vvalue;
} RUBY_PARAMETER;

#define rb_define_singleton_alias(klass,new,old) rb_define_alias(rb_singleton_class(klass),new,old)


/*
 * Function declarations
 */
VALUE ii_execute (int param_argc, VALUE * param_argv, VALUE param_self);
VALUE ii_connect (int param_argc, VALUE * param_argv, VALUE param_self);
static VALUE ing_disconnect (VALUE param_self);
VALUE ing_connect (VALUE param_self, VALUE param_targetDB);
static void ii_conn_init(II_CONN *ii_conn);
VALUE ing_init (int argc, VALUE *argv, VALUE self);
void ing_api_init ();
static VALUE rb_ingres_alloc(VALUE klass);
static void free_ii_conn (II_CONN *ing_conn);
static void ing_starttransaction();
static void ing_commit();
static void ing_rollback();
VALUE ing_query (char *param_sqlText);
static int ing_querytype(char *queryText);
VALUE ing_execute (VALUE argc, VALUE *argv, VALUE self);
II_PTR executeQuery (II_CONN *ii_conn, char *param_sqlText);
static void ing_conn_init(II_CONN *ing_conn);
void ii_api_query_close (II_CONN *ii_conn);
void ii_api_connect (II_CONN *ii_conn, char *param_targetDB, char *param_username, char *param_password);
void ii_api_commit (II_CONN *ii_conn);
static int ii_query_type(char *queryText);
void ii_api_rollback (II_CONN *ii_conn, II_SAVEPOINT_ENTRY *savePtEntry);
void ii_api_disconnect( II_CONN *ii_conn);
void ii_api_savepoint (II_CONN *ii_conn, VALUE savePtName);
void ii_free_savePtEntries (II_SAVEPOINT_ENTRY *savePtEntry);

void ii_api_set_connect_param (II_CONN *ii_conn, II_LONG paramID, VALUE paramValue);
void ii_api_set_env_param (II_LONG paramID, VALUE paramValue);

/* Transaction control */
VALUE ii_commit (VALUE param_self);
VALUE ii_rollback (int param_argc, VALUE * param_argv, VALUE param_self);
VALUE ii_savepoint (VALUE param_self, VALUE param_savepointName);
VALUE ii_savepoint_name(II_CONN *ii_conn, char *statement);

/* Memory Allocation/Deallocation */
void *ii_allocate (size_t nitems, size_t size);
void *ii_reallocate (void *oldPtr, size_t nitems, size_t size);
void ii_free (void **ptr);

/* TODO - The following has been taken from the Ingres CL and should be removed/replaced at some point */
# define        NULLCHAR        ('\0')	/* string terminator */
# define        EOS             NULLCHAR
# define        CM_A_DBL1       0x80	/* 1st byte of double byte character */
# define        CM_A_SPACE      0x10	/* SPACE */
# define        cmdblspace(str) (((*(str)&0377) == 0xA1) && ((*((str)+1)&0377) == 0xA1))

# if defined(NT_GENERIC)
FUNC_EXTERN u_i2 *CMgetAttrTab ();
FUNC_EXTERN char *CMgetCaseTab ();
# endif

# define CMbytecnt(ptr)         (1)
# define CMnext(str)            (++(str))
# define CMdbl1st(str)          (FALSE)

# if defined(_MSC_VER) || (defined(NT_GENERIC) && defined(IMPORT_DLL_DATA))
# define CMwhite(str)           ((((u_i2 *)CMgetAttrTab())[*(str)&0377] & CM_A_SPACE) != 0)
# else /* NT_GENERIC && IMPORT_DLL_DATA */
# define CMwhite(str)           ((CM_AttrTab[*(str)&0377] & CM_A_SPACE) != 0)
# endif	/* NT_GENERIC && IMPORT_DLL_DATA */

/* available via Ingres Compatibility Layer */
/* TODO - Remove dependency on the Ingres CL */
size_t STtrmwhite (register char *string);

/* Certain Ruby releases shipped with different Linux distributions do not contain
 * the necessary defines. E.g. RedHat ES 5.4 ships with Ruby 1.8.5.
 */
#ifdef RUBY_VERSION_CODE
#  if RUBY_VERSION_CODE < 186
#    define RARRAY_LEN(s) (RARRAY(s)->len)
#    define RSTRING_PTR(s) (RSTRING(s)->ptr)
#    define RSTRING_LEN(s) (RSTRING(s)->len)
#  endif /* RUBY_VERSION_CODE < 186 */
#endif

/*
vim:  ts=2 sw=2 expandtab
*/
